File: src/main/java/com/yupi/airouter/controller/ApiKeyController.java
package com.yupi.airouter.controller;

import cn.hutool.core.bean.BeanUtil;
import com.mybatisflex.core.paginate.Page;
import com.mybatisflex.core.query.QueryWrapper;
import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.DeleteRequest;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.dto.apikey.ApiKeyCreateRequest;
import com.yupi.airouter.model.entity.ApiKey;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.model.vo.ApiKeyVO;
import com.yupi.airouter.service.ApiKeyService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

/**
 * API Key 接口
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/api/key")
@Slf4j
public class ApiKeyController {

    @Resource
    private ApiKeyService apiKeyService;

    @Resource
    private UserService userService;

    /**
     * 创建 API Key
     */
    @PostMapping("/create")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "创建 API Key")
    public BaseResponse<ApiKeyVO> createApiKey(@RequestBody ApiKeyCreateRequest request,
                                                HttpServletRequest httpRequest) {
        User loginUser = userService.getLoginUser(httpRequest);
        ApiKey apiKey = apiKeyService.createApiKey(request.getKeyName(), loginUser);

        // 转换为 VO（完整显示 Key 值）
        ApiKeyVO apiKeyVO = BeanUtil.copyProperties(apiKey, ApiKeyVO.class);

        return ResultUtils.success(apiKeyVO);
    }

    /**
     * 获取我的 API Key 列表（分页）
     */
    @GetMapping("/list/my")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的 API Key 列表")
    public BaseResponse<Page<ApiKeyVO>> listMyApiKeys(
            @RequestParam(defaultValue = "1") long pageNum,
            @RequestParam(defaultValue = "10") long pageSize,
            HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);

        // 分页查询
        QueryWrapper queryWrapper = QueryWrapper.create()
                .eq("userId", loginUser.getId())
                .eq("isDelete", 0)
                .orderBy("createTime", false);

        Page<ApiKey> apiKeyPage = apiKeyService.page(Page.of(pageNum, pageSize), queryWrapper);

        // 转换为 VO（列表中部分隐藏 Key 值）
        Page<ApiKeyVO> apiKeyVOPage = new Page<>(pageNum, pageSize, apiKeyPage.getTotalRow());
        List<ApiKeyVO> apiKeyVOList = apiKeyPage.getRecords().stream()
                .map(apiKey -> {
                    ApiKeyVO vo = BeanUtil.copyProperties(apiKey, ApiKeyVO.class);
                    // 隐藏部分 Key 值（只显示前8位和后4位）
                    if (vo.getKeyValue() != null && vo.getKeyValue().length() > 12) {
                        String key = vo.getKeyValue();
                        vo.setKeyValue(key.substring(0, 8) + "****" + key.substring(key.length() - 4));
                    }
                    return vo;
                })
                .collect(Collectors.toList());

        apiKeyVOPage.setRecords(apiKeyVOList);
        return ResultUtils.success(apiKeyVOPage);
    }

    /**
     * 撤销 API Key
     */
    @PostMapping("/revoke")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "撤销 API Key")
    public BaseResponse<Boolean> revokeApiKey(@RequestBody DeleteRequest deleteRequest,
                                               HttpServletRequest request) {
        if (deleteRequest == null || deleteRequest.getId() == null) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }

        User loginUser = userService.getLoginUser(request);
        boolean result = apiKeyService.revokeApiKey(deleteRequest.getId(), loginUser.getId());

        return ResultUtils.success(result);
    }
}

File: go-backend/internal/controller/apikey_controller.go
package controller

import (
	"log"
	"strconv"

	"github.com/gin-gonic/gin"

	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/model/vo"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type ApiKeyController struct {
	apiKeyService *service.ApiKeyService
	userService   *service.UserService
}

func NewApiKeyController(apiKeyService *service.ApiKeyService, userService *service.UserService) *ApiKeyController {
	return &ApiKeyController{
		apiKeyService: apiKeyService,
		userService:   userService,
	}
}

func (c *ApiKeyController) CreateApiKey(ctx *gin.Context) {
	var request dto.ApiKeyCreateRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("create api key bind failed: %v", err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	apiKey, err := c.apiKeyService.CreateApiKey(request.KeyName, loginUser)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	common.Success(ctx, vo.ApiKeyVO{
		ID:           apiKey.ID,
		KeyValue:     apiKey.KeyValue,
		KeyName:      apiKey.KeyName,
		Status:       apiKey.Status,
		TotalTokens:  apiKey.TotalTokens,
		LastUsedTime: apiKey.LastUsedTime,
		CreateTime:   apiKey.CreateTime,
	})
}

func (c *ApiKeyController) ListMyApiKeys(ctx *gin.Context) {
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	pageNum := parsePositiveInt64WithDefault(ctx.Query("pageNum"), 1)
	pageSize := parsePositiveInt64WithDefault(ctx.Query("pageSize"), 10)
	pageData, err := c.apiKeyService.ListMyApiKeys(loginUser.ID, pageNum, pageSize)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	common.Success(ctx, pageData)
}

func (c *ApiKeyController) RevokeApiKey(ctx *gin.Context) {
	var request dto.DeleteRequest
	if err := ctx.ShouldBindJSON(&request); err != nil {
		log.Printf("revoke api key bind failed: %v", err)
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	if request.ID == nil || request.ID.Int64() <= 0 {
		common.Error(ctx, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	loginUser, err := c.userService.GetLoginUser(ctx)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	ok, err := c.apiKeyService.RevokeApiKey(request.ID.Int64(), loginUser.ID)
	if err != nil {
		c.handleError(ctx, err)
		return
	}
	common.Success(ctx, ok)
}

func (c *ApiKeyController) handleError(ctx *gin.Context, err error) {
	if bizErr, ok := errno.AsBusinessError(err); ok {
		common.Error(ctx, bizErr.Code, bizErr.Message)
		return
	}
	common.Error(ctx, errno.SystemError.Code, "系统错误")
}

func parsePositiveInt64WithDefault(raw string, defaultValue int64) int64 {
	value, err := strconv.ParseInt(raw, 10, 64)
	if err != nil || value <= 0 {
		return defaultValue
	}
	return value
}
